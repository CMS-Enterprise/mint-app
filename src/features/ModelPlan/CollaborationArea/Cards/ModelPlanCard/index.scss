@use 'uswds-core' as *;

.card--model-plan {
  min-height: 420px;

  .card__section-status {
    display: flex;
    gap: 24px;
    padding: 0 24px;
    margin-block: 16px;
  }
}
