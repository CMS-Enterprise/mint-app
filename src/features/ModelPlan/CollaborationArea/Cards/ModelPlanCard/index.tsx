import React, { useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  Grid
} from '@trussworks/react-uswds';
import {
  getITSolutionsStatus,
  StatusMessageType
} from 'features/ModelPlan/TaskList';
import { TaskListStatusTag } from 'features/ModelPlan/TaskList/_components/TaskListItem';
import { TaskStatus, useGetModelPlanQuery } from 'gql/generated/graphql';

import { Avatar } from 'components/Avatar';
import UswdsReactLink from 'components/LinkWrapper';
import Modal from 'components/Modal';
import ShareExportModal from 'components/ShareExport';
import Spinner from 'components/Spinner';
import { formatDateLocal } from 'utils/date';

// importing global card styles from Cards/cards.scss
import '../cards.scss';

type ModelPlanCardType = {
  modelID: string;
  setStatusMessage: (message: StatusMessageType) => void;
};

const ModelPlanCard = ({ modelID, setStatusMessage }: ModelPlanCardType) => {
  const { t: collaborationAreaT } = useTranslation('collaborationArea');
  const [isExportModalOpen, setIsExportModalOpen] = useState<boolean>(false);
  const { data, loading } = useGetModelPlanQuery({
    variables: {
      id: modelID
    }
  });

  const modelPlan = data?.modelPlan;

  // Returns the number of sections that have been started (i.e. not in 'READY' status)
  const sectionStartedCounter = useMemo(() => {
    if (loading || !modelPlan) return 0;

    const sections = [
      modelPlan.basics.status,
      modelPlan.generalCharacteristics.status,
      modelPlan.participantsAndProviders.status,
      modelPlan.beneficiaries.status,
      modelPlan.opsEvalAndLearning.status,
      modelPlan.payments.status,
      getITSolutionsStatus(modelPlan.operationalNeeds)
    ];

    return sections.filter(status => status !== TaskStatus.READY).length;
  }, [loading, modelPlan]);

  if (loading && !modelPlan)
    return (
      <Grid
        desktop={{ col: 6 }}
        className="padding-1 display-flex flex-column flex-align-center flex-justify-center height-mobile"
      >
        <Spinner />
      </Grid>
    );

  if (!modelPlan) return null;

  const { modifiedDts, modifiedByUserAccount, taskListStatus } = modelPlan;

  return (
    <>
      <Modal
        isOpen={isExportModalOpen}
        closeModal={() => setIsExportModalOpen(false)}
        className="padding-0 radius-md share-export-modal__container"
        navigation
        shouldCloseOnOverlayClick
      >
        <ShareExportModal
          closeModal={() => setIsExportModalOpen(false)}
          modelID={modelID}
          setStatusMessage={setStatusMessage}
        />
      </Modal>
      <Card
        gridLayout={{ desktop: { col: 6 } }}
        className="collaboration-area__card collaboration-area__main-card"
      >
        <CardHeader>
          <h3 className="usa-card__heading">
            {collaborationAreaT('modelPlanCard.heading')}
          </h3>
        </CardHeader>
        <div className="collaboration-area__status flex-align-center">
          <TaskListStatusTag
            status={taskListStatus}
            classname="width-fit-content"
          />
          <span className="text-base">
            {collaborationAreaT('modelPlanCard.sectionsStarted', {
              sectionsStarted: sectionStartedCounter
            })}
          </span>
        </div>

        <CardBody>
          <p>{collaborationAreaT('modelPlanCard.body')}</p>

          {modifiedDts && modifiedByUserAccount && (
            <div className="display-inline tablet:display-flex margin-top-2 margin-bottom-3 flex-align-center">
              <span className="text-base margin-right-1">
                {collaborationAreaT('modelPlanCard.mostRecentEdit', {
                  date: formatDateLocal(modifiedDts, 'MM/dd/yyyy')
                })}
              </span>
              <Avatar
                className="text-base-darkest"
                user={modifiedByUserAccount.commonName}
              />
            </div>
          )}
        </CardBody>

        <CardFooter>
          <UswdsReactLink
            to={`/models/${modelID}/task-list`}
            className="usa-button margin-right-1"
            variant="unstyled"
            data-testid="to-task-list"
          >
            {collaborationAreaT('goToModelPlan')}
          </UswdsReactLink>
          {sectionStartedCounter !== 0 && (
            <Button
              type="button"
              className="usa-button usa-button--outline"
              onClick={() => setIsExportModalOpen(true)}
            >
              {collaborationAreaT('modelPlanCard.shareButton')}
            </Button>
          )}
        </CardFooter>
      </Card>
    </>
  );
};

export default ModelPlanCard;
