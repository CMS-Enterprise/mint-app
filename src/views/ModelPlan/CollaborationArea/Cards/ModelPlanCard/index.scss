.card--model-plan {
  min-height: 420px;
}
