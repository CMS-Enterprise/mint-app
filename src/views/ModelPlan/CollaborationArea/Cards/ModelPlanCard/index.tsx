import React, { useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  Grid
} from '@trussworks/react-uswds';
import { TaskStatus, useGetModelPlanQuery } from 'gql/gen/graphql';

import UswdsReactLink from 'components/LinkWrapper';
import Modal from 'components/Modal';
import { Avatar } from 'components/shared/Avatar';
import ShareExportModal from 'components/ShareExport';
import Spinner from 'components/Spinner';
import { formatDateLocal } from 'utils/date';
import {
  getITSolutionsStatus,
  StatusMessageType
} from 'views/ModelPlan/TaskList';
import { TaskListStatusTag } from 'views/ModelPlan/TaskList/_components/TaskListItem';

// importing global card styles from Cards/cards.scss
import '../cards.scss';
// import './index.scss';
import './index.scss';

type ModelPlanCardType = {
  modelID: string;
  setStatusMessage: (message: StatusMessageType) => void;
};

const ModelPlanCard = ({ modelID, setStatusMessage }: ModelPlanCardType) => {
  const { t: collaborationAreaT } = useTranslation('collaborationArea');
  const [isExportModalOpen, setIsExportModalOpen] = useState<boolean>(false);
  const { data, loading } = useGetModelPlanQuery({
    variables: {
      id: modelID
    }
  });

  const modelPlan = data?.modelPlan;

  // Returns the number of sections that have been started (i.e. not in 'READY' status)
  const sectionStartedCounter = useMemo(() => {
    if (loading || !modelPlan) return 0;

    const sections = [
      modelPlan.basics.status,
      modelPlan.generalCharacteristics.status,
      modelPlan.participantsAndProviders.status,
      modelPlan.beneficiaries.status,
      modelPlan.opsEvalAndLearning.status,
      modelPlan.payments.status,
      getITSolutionsStatus(modelPlan.operationalNeeds)
    ];

    return sections.filter(status => status !== TaskStatus.READY).length;
  }, [loading, modelPlan]);

  if (loading && !modelPlan)
    return (
      <Grid
        desktop={{ col: 6 }}
        className="padding-1 display-flex flex-column flex-align-center flex-justify-center height-mobile"
      >
        <Spinner />
      </Grid>
    );

  if (!modelPlan) return null;

  const { modifiedDts, modifiedByUserAccount, taskListStatus } = modelPlan;

  return (
    <>
      <Modal
        isOpen={isExportModalOpen}
        closeModal={() => setIsExportModalOpen(false)}
        className="padding-0 radius-md share-export-modal__container"
        navigation
        shouldCloseOnOverlayClick
      >
        <ShareExportModal
          closeModal={() => setIsExportModalOpen(false)}
          modelID={modelID}
          setStatusMessage={setStatusMessage}
        />
      </Modal>
      <Card
        gridLayout={{ desktop: { col: 6 } }}
        className="collaboration-area__card card--model-plan"
      >
        <CardHeader>
          <h3 className="usa-card__heading">
            {collaborationAreaT('modelPlanCard.heading')}
          </h3>
        </CardHeader>
        <div className="card__section-status flex-align-center">
          <TaskListStatusTag
            status={taskListStatus}
            classname="width-fit-content"
          />
          <span className="text-base">
            {collaborationAreaT('modelPlanCard.sectionsStarted', {
              sectionsStarted: sectionStartedCounter
            })}
          </span>
        </div>

        <CardBody>
          <p>{collaborationAreaT('modelPlanCard.body')}</p>
        </CardBody>

        {modifiedDts && modifiedByUserAccount && (
          <div className="display-inline tablet:display-flex margin-top-2 margin-bottom-3 flex-align-center padding-x-3">
            <span className="text-base margin-right-1">
              {collaborationAreaT('modelPlanCard.mostRecentEdit', {
                date: formatDateLocal(modifiedDts, 'MM/dd/yyyy')
              })}
            </span>
            <Avatar
              className="text-base-darkest"
              user={modifiedByUserAccount.commonName}
            />
          </div>
        )}
        <CardFooter>
          <UswdsReactLink
            to={`/models/${modelID}/task-list`}
            className="usa-button"
            variant="unstyled"
            data-testid="to-task-list"
          >
            {collaborationAreaT('goToModelPlan')}
          </UswdsReactLink>
          {sectionStartedCounter !== 0 && (
            <Button
              type="button"
              className="usa-button usa-button--outline"
              onClick={() => setIsExportModalOpen(true)}
            >
              {collaborationAreaT('modelPlanCard.shareButton')}
            </Button>
          )}
        </CardFooter>
      </Card>
    </>
  );
};

export default ModelPlanCard;
