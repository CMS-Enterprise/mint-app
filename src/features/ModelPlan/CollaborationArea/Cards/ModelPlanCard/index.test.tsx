import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { waitFor } from '@testing-library/react';
import {
  GetModelPlanDocument,
  GetModelPlanQuery,
  ModelPhase,
  ModelStatus,
  PrepareForClearanceStatus,
  TaskStatus
} from 'gql/generated/graphql';
import VerboseMockedProvider from 'tests/MockedProvider';
import setup from 'tests/util';

import ModelPlanCard, { getLastModifiedSection } from './index';

const modelID: string = 'f11eb129-2c80-4080-9440-439cbe1a286f';
type GetModelPlanTypes = GetModelPlanQuery['modelPlan'];

const modelPlan = {
  __typename: 'ModelPlan',
  isFavorite: true,
  id: '6e224030-09d5-46f7-ad04-4bb851b36eab',
  status: ModelStatus.PLAN_DRAFT,
  taskListStatus: TaskStatus.IN_PROGRESS,
  modelName: 'Test',
  opSolutionLastModifiedDts: '2022-05-12T15:01:39.190679Z',
  modifiedDts: '2022-05-12T15:01:39.190679Z',
  modifiedByUserAccount: {
    __typename: 'UserAccount',
    commonName: 'John Doe'
  },
  archived: false,
  suggestedPhase: {
    __typename: 'PhaseSuggestion',
    phase: ModelPhase.ICIP_COMPLETE,
    suggestedStatuses: [ModelStatus.ICIP_COMPLETE]
  },
  basics: {
    __typename: 'PlanBasics',
    id: '123',
    modifiedDts: null,
    clearanceStarts: '2022-05-12T15:01:39.190679Z',
    readyForClearanceDts: '2022-05-12T15:01:39.190679Z',
    status: TaskStatus.READY
  },
  opsEvalAndLearning: {
    __typename: 'PlanOpsEvalAndLearning',
    id: '7865676',
    createdBy: 'John Doe',
    createdDts: '',
    modifiedBy: '',
    modifiedDts: '',
    readyForClearanceDts: '',
    status: TaskStatus.IN_PROGRESS
  },
  generalCharacteristics: {
    __typename: 'PlanGeneralCharacteristics',
    id: '54234',
    createdBy: 'John Doe',
    createdDts: '',
    modifiedBy: '',
    modifiedDts: '',
    readyForClearanceDts: '',
    status: TaskStatus.IN_PROGRESS
  },
  participantsAndProviders: {
    __typename: 'PlanParticipantsAndProviders',
    id: '46246356',
    createdBy: 'John Doe',
    createdDts: '',
    modifiedBy: '',
    modifiedDts: '',
    readyForClearanceDts: '',
    status: TaskStatus.IN_PROGRESS
  },
  beneficiaries: {
    __typename: 'PlanBeneficiaries',
    id: '09865643',
    createdBy: 'John Doe',
    createdDts: '',
    modifiedBy: '',
    modifiedDts: '',
    readyForClearanceDts: '',
    status: TaskStatus.IN_PROGRESS
  },
  prepareForClearance: {
    __typename: 'PrepareForClearance',
    status: PrepareForClearanceStatus.IN_PROGRESS,
    modifiedDts: ''
  },
  payments: {
    __typename: 'PlanPayments',
    id: '8756435235',
    createdBy: 'John Doe',
    createdDts: '',
    modifiedBy: '',
    modifiedDts: '',
    readyForClearanceDts: '',
    status: TaskStatus.IN_PROGRESS
  },
  crs: [],
  tdls: [],
  echimpCRsAndTDLs: [],
  operationalNeeds: [] as any,
  documents: [
    {
      __typename: 'PlanDocument',
      id: '6e224030-09d5-46f7-ad04-4bb851b36eab',
      fileName: 'test.pdf',
      fileType: 'application/pdf'
    }
  ],
  collaborators: [],
  discussions: [
    {
      __typename: 'PlanDiscussion',
      id: '123',
      content: {
        __typename: 'TaggedContent',
        rawContent: 'This is a question.'
      },
      createdBy: 'John Doe',
      createdDts: '2022-05-12T15:01:39.190679Z',
      replies: []
    },
    {
      __typename: 'PlanDiscussion',
      id: '456',
      content: {
        __typename: 'TaggedHTML',
        rawContent: 'This is a second question.'
      },
      createdBy: 'Jane Doe',
      createdDts: '2022-05-12T15:01:39.190679Z',
      replies: [
        {
          __typename: 'DiscussionReply',
          discussionID: '456',
          id: 'abc',
          content: {
            __typename: 'TaggedHTML',
            rawContent: 'This is an answer.'
          },
          createdBy: 'Jack Doe',
          createdDts: '2022-05-12T15:01:39.190679Z'
        }
      ]
    }
  ]
} as GetModelPlanTypes;

const modelPlanMocks = [
  {
    request: {
      query: GetModelPlanDocument,
      variables: { id: modelID }
    },
    result: {
      data: {
        modelPlan
      }
    }
  }
];

describe('ModelPlanCard', () => {
  it('renders without issues', async () => {
    const { getByText } = setup(
      <MemoryRouter initialEntries={[`/models/${modelID}/collaboration-area`]}>
        <VerboseMockedProvider mocks={modelPlanMocks} addTypename={false}>
          <ModelPlanCard modelID={modelID} setStatusMessage={() => null} />
        </VerboseMockedProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(getByText('Model Plan')).toBeInTheDocument();
    });
  });

  it('returns null if no section of task list is started', () => {
    const result = getLastModifiedSection(modelPlan);
    expect(result).toBeUndefined();
  });

  it('returns most recent task list section', () => {
    const startedModelPlan = { ...modelPlan };

    startedModelPlan.beneficiaries.modifiedDts = '2021-05-12T15:01:39.190679Z';
    startedModelPlan.beneficiaries.modifiedByUserAccount = {
      __typename: 'UserAccount',
      commonName: 'Jane Doe'
    };

    startedModelPlan.generalCharacteristics.modifiedDts =
      '2022-05-12T15:01:39.190679Z';
    startedModelPlan.generalCharacteristics.modifiedByUserAccount = {
      __typename: 'UserAccount',
      commonName: 'John Doe'
    };

    const result = getLastModifiedSection(startedModelPlan);
    expect(result).toEqual(startedModelPlan.generalCharacteristics);
  });

  it('matches the snapshot', async () => {
    const { getByText, queryByText, asFragment } = setup(
      <MemoryRouter initialEntries={[`/models/${modelID}/collaboration-area`]}>
        <VerboseMockedProvider mocks={modelPlanMocks} addTypename={false}>
          <ModelPlanCard modelID={modelID} setStatusMessage={() => null} />
        </VerboseMockedProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(getByText('Model Plan')).toBeInTheDocument();
      expect(
        queryByText('Most recent edit on 05/12/2022 by')
      ).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
