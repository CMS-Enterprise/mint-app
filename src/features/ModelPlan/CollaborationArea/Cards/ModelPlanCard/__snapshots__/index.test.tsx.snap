// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelPlanCard > matches the snapshot 1`] = `
<DocumentFragment>
  <li
    class="usa-card desktop:grid-col-6 collaboration-area__card card--model-plan"
    data-testid="Card"
  >
    <div
      class="usa-card__container"
    >
      <header
        class="usa-card__header"
        data-testid="CardHeader"
      >
        <h3
          class="usa-card__heading"
        >
          Model Plan
        </h3>
      </header>
      <div
        class="card__section-status flex-align-center"
      >
        <div
          class="model-plan-task-list__task-tag line-height-body-1 text-bold mint-no-print bg-warning width-fit-content"
          data-testid="tasklist-tag"
        >
          <span>
            In progress
          </span>
        </div>
        <span
          class="text-base"
        >
          5/7 sections started
        </span>
      </div>
      <div
        class="usa-card__body"
        data-testid="CardBody"
      >
        <p>
          The Model Plan will help components across CMS evaluate your model’s operational requirements and IT needs. It contains questions about payments, providers, general characteristics, and more. The Model Plan is flexible, so you may leave questions blank, add new information, and change information as you iterate on your model or learn of new dependencies.
        </p>
      </div>
      <div
        class="display-inline tablet:display-flex margin-top-2 margin-bottom-3 flex-align-center padding-x-3"
      >
        <span
          class="text-base margin-right-1"
        >
          Most recent edit on 05/12/2022 by 
        </span>
        <div
          class="display-flex text-base-darkest flex-align-center"
        >
          <div
            class="display-flex flex-align-center flex-justify-center minw-4 circle-4 bg-blue-10 margin-right-1"
            data-testid="avatar--basic"
          >
            JD
          </div>
          <div
            class="margin-y-0"
          >
            <span
              class=""
            >
              John Doe
            </span>
          </div>
        </div>
      </div>
      <div
        class="usa-card__footer"
        data-testid="CardFooter"
      >
        <a
          class="usa-button margin-right-1"
          data-testid="to-task-list"
          href="/models/f11eb129-2c80-4080-9440-439cbe1a286f/task-list"
        >
          Go to Model Plan
        </a>
        <button
          class="usa-button usa-button usa-button--outline"
          data-testid="button"
          type="button"
        >
          Share or export Model Plan
        </button>
      </div>
    </div>
  </li>
</DocumentFragment>
`;

exports[`ModelPlanCard > matches the snapshot 2`] = `
<DocumentFragment>
  <li
    class="usa-card desktop:grid-col-6 collaboration-area__card collaboration-area__main-card"
    data-testid="Card"
  >
    <div
      class="usa-card__container"
    >
      <header
        class="usa-card__header"
        data-testid="CardHeader"
      >
        <h3
          class="usa-card__heading"
        >
          Model Plan
        </h3>
      </header>
      <div
        class="collaboration-area__status flex-align-center"
      >
        <div
          class="model-plan-task-list__task-tag line-height-body-1 text-bold mint-no-print bg-warning width-fit-content"
          data-testid="tasklist-tag"
        >
          <span>
            In progress
          </span>
        </div>
        <span
          class="text-base"
        >
          5/7 sections started
        </span>
      </div>
      <div
        class="usa-card__body"
        data-testid="CardBody"
      >
        <p>
          The Model Plan will help components across CMS evaluate your model’s operational requirements and IT needs. It contains questions about payments, providers, general characteristics, and more. The Model Plan is flexible, so you may leave questions blank, add new information, and change information as you iterate on your model or learn of new dependencies.
        </p>
        <div
          class="display-inline tablet:display-flex margin-top-2 margin-bottom-3 flex-align-center"
        >
          <span
            class="text-base margin-right-1"
          >
            Most recent edit on 05/12/2022 by 
          </span>
          <div
            class="display-flex text-base-darkest flex-align-center"
          >
            <div
              class="display-flex flex-align-center flex-justify-center minw-4 circle-4 bg-blue-10 margin-right-1"
              data-testid="avatar--basic"
            >
              JD
            </div>
            <div
              class="margin-y-0"
            >
              <span
                class=""
              >
                John Doe
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="usa-card__footer"
        data-testid="CardFooter"
      >
        <a
          class="usa-button margin-right-1"
          data-testid="to-task-list"
          href="/models/f11eb129-2c80-4080-9440-439cbe1a286f/task-list"
        >
          Go to Model Plan
        </a>
        <button
          class="usa-button usa-button usa-button--outline"
          data-testid="button"
          type="button"
        >
          Share or export Model Plan
        </button>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
